File: app/src/main/java/com/sofps/inspirationalquotes/data/source/local/QuotesLocalDataSource.java
package com.sofps.inspirationalquotes.data.source.local;

import android.support.annotation.NonNull;
import com.sofps.inspirationalquotes.data.DataBaseHelper;
import com.sofps.inspirationalquotes.data.Quote;
import com.sofps.inspirationalquotes.data.source.QuotesDataSource;
import java.util.ArrayList;
import java.util.List;

public class QuotesLocalDataSource implements QuotesDataSource {

    private final DataBaseHelper mDataBaseHelper;

    public QuotesLocalDataSource(DataBaseHelper dataBaseHelper) {
        mDataBaseHelper = dataBaseHelper;
    }

    @Override
    public void getQuote(@NonNull String language, @NonNull GetQuoteCallback callback) {
        callback.onQuoteLoaded(getQuotes(language).get(0)); // TODO
    }

    public void persist(Quote quote) {
        mDataBaseHelper.insertQuote(quote);
    }

    public void addOneTimeShowed(Quote quote) {
        mDataBaseHelper.addOneTimeShowed(quote);
    }

    private List<Quote> getQuotes(String language) {
        DataBaseHelper.QuoteCursor cursor = mDataBaseHelper.queryQuotes(language);
        List<Quote> quotes = new ArrayList<>();
        while (cursor.moveToNext()) {
            quotes.add(cursor.getQuote());
        }
        cursor.close();
        mDataBaseHelper.close();
        return quotes;
    }
}

File: app/src/main/java/com/sofps/inspirationalquotes/data/source/local/QuotesLocalDataSource.kt
package com.sofps.inspirationalquotes.data.source.local

import com.sofps.inspirationalquotes.data.source.QuotesDataSource
import com.sofps.inspirationalquotes.data.QuoteDb
import com.sofps.inspirationalquotes.data.QuoteDao
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.launch

class QuotesLocalDataSource(
        private val quoteDao: QuoteDao
) : QuotesDataSource {

    override fun getQuotes(language: String): Flow<List<QuoteDb>> {
        return flow {
            val quotes = quoteDao.getAllByLanguage(language)
            emit(quotes)
        }.flowOn(Dispatchers.IO)
    }

    fun persist(quotes: List<QuoteDb>) {
        GlobalScope.launch(Dispatchers.IO) {
            quoteDao.insertAll(quotes)
        }
    }

    fun addOneTimeShowed(quote: QuoteDb) {
        val updatedQuote = quote.copy(timesShowed = quote.timesShowed + 1)

        GlobalScope.launch(Dispatchers.IO) {
            quoteDao.update(updatedQuote)
        }
    }
}
